pkg/kubectl/cmd/drain.go

```diff
@@ -0,0 +1,163 @@
+		Example: drain_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(
+				RunDrain(f, out, cmd, args, options))
+		},
+	}
+	cmd.Flags().BoolVar(&options.Force, "force", false, "Delete unreplicated pods.")
+	cmd.Flags().IntVar(&options.GracePeriodSeconds, "grace-period", -1, "Period of time in seconds given to each pod to terminate gracefully. Ignored if negative.")
+	return cmd
+}
+
+func RunDrain(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string, options *DrainOptions) error {
+	if len(args) != 1 {
+		return cmdutil.UsageError(cmd, fmt.Sprintf("USAGE: %s [flags]", cmd.Use))
+	}
+	options.NodeName = args[0]
+
+	mapper, _ := f.Object()
+	c, err := f.Client()
+	if err != nil {
+		return err
+	}
+	if err = maybeLameNode(mapper, c, out, options); err != nil {
+		return err
+	}
+	pods, err := c.Pods(api.NamespaceAll).List(nil, fields.Set{"spec.nodeName": options.NodeName}.AsSelector())
+	if err != nil {
+		return err
+	}
+	if err = scanForReplicationControllers(c, out, options, pods); err != nil {
+		return err
+	}
+	if err = deletePods(mapper, c, out, options, pods); err != nil {
+		return err
+	}
+	cmdutil.PrintSuccess(mapper, false, out, "node", options.NodeName, "drained")
+	return nil
+}
+
+func maybeLameNode(mapper meta.RESTMapper, c *client.Client, out io.Writer, options *DrainOptions) error {
+	node, err := c.Nodes().Get(options.NodeName)
+	if err != nil {
+		return err
+	}
+	if node.Spec.Unschedulable {
+		cmdutil.PrintSuccess(mapper, false, out, "node", node.Name, "already unschedulable")
+	} else {
+		node.Spec.Unschedulable = true
+		newNode, err := c.Nodes().Update(node)
+		if err != nil {
+			return err
+		}
+		cmdutil.PrintSuccess(mapper, false, out, "node", newNode.Name, "marked unschedulable")
+	}
+	return nil
+}
+
+func scanForReplicationControllers(c *client.Client, out io.Writer, options *DrainOptions, pods *api.PodList) error {
+	rcs, err := c.ReplicationControllers(api.NamespaceAll).List(nil, nil)
+	if err != nil {
+		return err
+	}
+	unreplicatedPodNames := []string{}
+
+	for _, pod := range pods.Items {
+		replicated := false
+		for _, rc := range rcs.Items {
+			selector := labels.SelectorFromSet(rc.Spec.Selector)
+			if selector.Matches(labels.Set(pod.Labels)) {
+				replicated = true
+				break
+			}
+		}
+		if !replicated {
+			unreplicatedPodNames = append(unreplicatedPodNames, pod.Name)
+		}
+	}
+	if len(unreplicatedPodNames) > 0 {
+		joined := strings.Join(unreplicatedPodNames, ", ")
+		if !options.Force {
+			return fmt.Errorf("Refusing to continue due to unreplicated pods: %s (use --force to override)", joined)
+		}
+		fmt.Fprintf(out, "WARNING: Deleting these unreplicated pods: %s\n", joined)
+	}
+	return nil
+}
+
+func deletePods(mapper meta.RESTMapper, c *client.Client, out io.Writer, options *DrainOptions, pods *api.PodList) error {
+	gracePeriodSeconds := int64(options.GracePeriodSeconds)
+	deleteOptions := api.DeleteOptions{GracePeriodSeconds: &gracePeriodSeconds}
```

Comment:
I think there are a couple of issues here
* You don't sanitize the value given on the command line. Someone could specify a negative number (in fact the default is a negative number).
* I think we want the Borg behavior where if the user doesn't specify grace period on the command line, we use the grace period specified in the pod (if any). You can get this by just doing api.DeleteOptions{} . So the specific changs I would make are: (1) update the help text to indicate that if you don't specify a grace period on the command line, then the pod-specified grace period will be used (if any); (2) have code here do `deleteOptions = api.DeleteOptions{}` if the gracePeriodSeconds is -1; (3) do something reasonable with gracePeriodSeconds that is less than -1.